fix: address golangci-lint warnings

andrewkroh opened the pull request:
This PR enables current golangci-lint checks and fixes the resulting issues.
- Pins golangci-lint v2.12.2 and go-licenser v0.4.2 while keeping tool dependencies out of the application module.
- Aligns the GitHub Actions golangci-lint version with local development.
- Adds package and exported symbol documentation.
- Checks previously ignored errors from `io.Copy` and `flag.Value.Set`.
- Replaces deprecated `io/ioutil` and `zap.OnFatal` APIs.
- Prevents network outputs from modifying caller-owned buffers.
- Propagates command cancellation through Pub/Sub setup.
- Uses caller context for GCS initialization while allowing pending writes to finish during close.
- Makes GCS cleanup safe after failed or skipped connection setup.
- Places `context.Context` first and replaces unused parameters with blank identifiers.
- Applies gofumpt formatting.
- Disables staticcheck QF1008 because explicit embedded field selectors improve clarity.

Validation:
- `make check-fmt`
- `make lint`
- Changed package tests and Docker-backed package test compilation

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/stream/builds/320)
* Commit: 7b2f37a2a8af2b224136d126e7dfb90dbcea2fad

### Failed CI Steps
* [:golangci-lint: Lint](https://buildkite.com/elastic/stream/builds/320#0197f15c-f1ad-4748-8e27-2e3df5df9849)

<!--buildkite-pr-comment-stream
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/320","number":320,"commit":"7b2f37a2a8af2b224136d126e7dfb90dbcea2fad"}],"number":320}
buildkite-pr-comment-->

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/stream/builds/321)
* Commit: e1201a2d94e3b011c7c2b22a155a36ce8d37887b

### Failed CI Steps
* [:golangci-lint: Lint](https://buildkite.com/elastic/stream/builds/321#0197f162-47be-47b4-962b-a3136b8b3a8e)

### History
* :broken_heart: [Build #320](https://buildkite.com/elastic/stream/builds/320) failed 7b2f37a2a8af2b224136d126e7dfb90dbcea2fad

<!--buildkite-pr-comment-stream
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/321","number":321,"commit":"e1201a2d94e3b011c7c2b22a155a36ce8d37887b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/320","number":320,"commit":"7b2f37a2a8af2b224136d126e7dfb90dbcea2fad"}],"number":321}
buildkite-pr-comment-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/stream/builds/323)
* Commit: 1230f427967d993da1edd5f33258d233af9e6713

### History
* :broken_heart: [Build #321](https://buildkite.com/elastic/stream/builds/321) failed e1201a2d94e3b011c7c2b22a155a36ce8d37887b
* :broken_heart: [Build #320](https://buildkite.com/elastic/stream/builds/320) failed 7b2f37a2a8af2b224136d126e7dfb90dbcea2fad

<!--buildkite-pr-comment-stream
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/323","number":323,"commit":"1230f427967d993da1edd5f33258d233af9e6713"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/321","number":321,"commit":"e1201a2d94e3b011c7c2b22a155a36ce8d37887b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/320","number":320,"commit":"7b2f37a2a8af2b224136d126e7dfb90dbcea2fad"}],"number":323}
buildkite-pr-comment-->

andrewkroh committed:
golangci.yml - ignore staticcheck QF1008

I think this reduces clarity when there are multiple embedded structs.

andrewkroh committed:
go.mod - add tool directives

Replace tools.go with go mod tool directives.

go get -tool github.com/golangci/golangci-lint/v2/cmd/golangci-lint
go get -tool github.com/elastic/go-licenser
go mod tidy

andrewkroh committed:
add godoc to packages and exported types

Rename GcsOptions to GCSOptions.

andrewkroh committed:
use blank identifier to unused parameters

andrewkroh committed:
make context.Context the first func parameter

andrewkroh committed:
pass cmd context to GCP pub/sub API calls

andrewkroh committed:
log errors from flag.Value.Set

andrewkroh committed:
do no modify data based into a io.Writer

Fixed an issue in TCP and TLS outputs where the write buffer was being modified, which is unsafe. A new buffer is now used.

andrewkroh committed:
gofumpt

andrewkroh committed:
replaced deprecated io/ioutil with io and os packages

andrewkroh committed:
replaced deprecated zap.OnFatal with zap.WithFatalHook

andrewkroh committed:
fix(output/gcs): context leak

Resolved a potential context leak in the GCS output by ensuring the cancel function is always called.

andrewkroh committed:
fix(webhook): add io.Copy error check

andrewkroh committed:
fix(tls,udp): match package name to dir name

andrewkroh committed:
add changelog

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/stream/builds/335)
* Commit: ba94778312e4a9de6bc9b379fa72ce16c8100a81

### History
* :green_heart: [Build #324](https://buildkite.com/elastic/stream/builds/324) succeeded 8523016267cda4a9077b3c49a246ed38c6fe2d2c
* :green_heart: [Build #323](https://buildkite.com/elastic/stream/builds/323) succeeded 1230f427967d993da1edd5f33258d233af9e6713
* :broken_heart: [Build #321](https://buildkite.com/elastic/stream/builds/321) failed e1201a2d94e3b011c7c2b22a155a36ce8d37887b
* :broken_heart: [Build #320](https://buildkite.com/elastic/stream/builds/320) failed 7b2f37a2a8af2b224136d126e7dfb90dbcea2fad

<!--buildkite-pr-comment-stream
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/335","number":335,"commit":"ba94778312e4a9de6bc9b379fa72ce16c8100a81"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/324","number":324,"commit":"8523016267cda4a9077b3c49a246ed38c6fe2d2c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/323","number":323,"commit":"1230f427967d993da1edd5f33258d233af9e6713"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/321","number":321,"commit":"e1201a2d94e3b011c7c2b22a155a36ce8d37887b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/320","number":320,"commit":"7b2f37a2a8af2b224136d126e7dfb90dbcea2fad"}],"number":335}
buildkite-pr-comment-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/stream/builds/340)
* Commit: b5a13b3c2afff02c03e1898249a6683d095c757f

### History
* :green_heart: [Build #335](https://buildkite.com/elastic/stream/builds/335) succeeded ba94778312e4a9de6bc9b379fa72ce16c8100a81
* :green_heart: [Build #324](https://buildkite.com/elastic/stream/builds/324) succeeded 8523016267cda4a9077b3c49a246ed38c6fe2d2c
* :green_heart: [Build #323](https://buildkite.com/elastic/stream/builds/323) succeeded 1230f427967d993da1edd5f33258d233af9e6713
* :broken_heart: [Build #321](https://buildkite.com/elastic/stream/builds/321) failed e1201a2d94e3b011c7c2b22a155a36ce8d37887b
* :broken_heart: [Build #320](https://buildkite.com/elastic/stream/builds/320) failed 7b2f37a2a8af2b224136d126e7dfb90dbcea2fad

<!--buildkite-pr-comment-stream
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/340","number":340,"commit":"b5a13b3c2afff02c03e1898249a6683d095c757f"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/335","number":335,"commit":"ba94778312e4a9de6bc9b379fa72ce16c8100a81"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/324","number":324,"commit":"8523016267cda4a9077b3c49a246ed38c6fe2d2c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/323","number":323,"commit":"1230f427967d993da1edd5f33258d233af9e6713"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/321","number":321,"commit":"e1201a2d94e3b011c7c2b22a155a36ce8d37887b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/320","number":320,"commit":"7b2f37a2a8af2b224136d126e7dfb90dbcea2fad"}],"number":340}
buildkite-pr-comment-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/stream/builds/358)
* Commit: dbd20271c63844086bf6d59798ea2960bf306677

### History
* :green_heart: [Build #340](https://buildkite.com/elastic/stream/builds/340) succeeded b5a13b3c2afff02c03e1898249a6683d095c757f
* :green_heart: [Build #335](https://buildkite.com/elastic/stream/builds/335) succeeded ba94778312e4a9de6bc9b379fa72ce16c8100a81
* :green_heart: [Build #324](https://buildkite.com/elastic/stream/builds/324) succeeded 8523016267cda4a9077b3c49a246ed38c6fe2d2c
* :green_heart: [Build #323](https://buildkite.com/elastic/stream/builds/323) succeeded 1230f427967d993da1edd5f33258d233af9e6713
* :broken_heart: [Build #321](https://buildkite.com/elastic/stream/builds/321) failed e1201a2d94e3b011c7c2b22a155a36ce8d37887b
* :broken_heart: [Build #320](https://buildkite.com/elastic/stream/builds/320) failed 7b2f37a2a8af2b224136d126e7dfb90dbcea2fad

<!--buildkite-pr-comment-stream
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/358","number":358,"commit":"dbd20271c63844086bf6d59798ea2960bf306677"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/340","number":340,"commit":"b5a13b3c2afff02c03e1898249a6683d095c757f"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/335","number":335,"commit":"ba94778312e4a9de6bc9b379fa72ce16c8100a81"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/324","number":324,"commit":"8523016267cda4a9077b3c49a246ed38c6fe2d2c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/323","number":323,"commit":"1230f427967d993da1edd5f33258d233af9e6713"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/stream/builds/321","number":321,"commit":"e1201a2d94e3b011c7c2b22a155a36ce8d37887b"}],"number":358}
buildkite-pr-comment-->

andrewkroh commented on go.mod:
```diff
@@ -3,114 +3,322 @@ module github.com/elastic/stream
+
+tool (
+	github.com/elastic/go-licenser
+	github.com/golangci/golangci-lint/v2/cmd/golangci-lint
```
Looking back at this months later, I'm less satisfied with the `go get -tool` changes. If others were consuming this project as a library I would be much less inclined to use this.

efd6 replied:
What is the concern?

efd6 commented on internal/output/options.go:
```diff
@@ -34,36 +36,47 @@ type WebhookOptions struct {
+// KafkaOptions holds configuration for the Kafka output.
 type KafkaOptions struct {
-	Topic string // Topic name. Will create it if not exists.
+	Topic string // Topic is the Kafka topic name. It will be created if it does not exist.
```
"Topic is the Kafka topic name. The topic will be created if it does not exist."

("It" is `Topic string` on first reading)

Similar below.

efd6 commented on internal/command/httpserver.go:
```diff
@@ -26,7 +26,7 @@ func newHTTPServerRunner(options *httpserver.Options, logger *zap.Logger) *cobra
 	}
 
-	r.cmd.RunE = func(_ *cobra.Command, args []string) error {
+	r.cmd.RunE = func(_ *cobra.Command, _ []string) error {
```
Just FYI, this could be `r.cmd.RunE = func(*cobra.Command, []string) error {`. No need for change.

efd6 commented on internal/output/tls/tls.go:
```diff
@@ -78,5 +78,13 @@ func (o *Output) Close() error {
+	buf := make([]byte, len(b)+1)
+	copy(buf, b)
+	buf[len(b)] = '\n'
+	return o.conn.Write(buf)
```
In both cases this could be `return o.conn.Write(append(b[:len(b):len(b)], '\n'))` with safety.

efd6 commented on internal/output/gcs/gcs.go:
```diff
@@ -95,6 +95,7 @@ func NewClient(addr string) (gcsClient *storage.Client, ctx context.Context, can
 	if addr != "" {
 		h, err = url.Parse(addr)
 		if err != nil {
+			cancel()
```
I think rather than doing this, `NewClient` should accept a `context.Context` and the caller should have the responsibility to handle this.

efd6 commented on internal/output/webhook/webhook.go:
```diff
@@ -121,7 +121,9 @@ func (o *Output) Write(b []byte) (int, error) {
+	if _, err = io.Copy(&buf, resp.Body); err != nil {
+		return 0, fmt.Errorf("failed to read response body: %w", err)
+	}
 	defer resp.Body.Close()
```
Before the copy.

andrewkroh committed:
Merge elastic/main into fix/linter-warnings

Resolve dependency conflicts and preserve safe network writes after the
TCP, TLS, and UDP outputs moved into the unified network output.

andrewkroh committed:
address review feedback

Clarify option docs, use caller-owned context for GCS, and simplify
resource-safe output handling.

andrewkroh committed:
fix(gcppubsub): propagate dial context

Allow cancellation to stop the emulator probe and resource cleanup instead
of using detached background contexts during setup.

andrewkroh committed:
build: pin Go tool versions

Run current fixed tool releases outside the application module so linting is
reproducible without adding tool dependencies to the runtime graph.
